test/test_fork.py

```diff
@@ -98,3 +107,53 @@ def test_topology_reset(self):
             self.assertNotEqual(child_id, init_id)
             passed, msg = parent_conn.recv()
             self.assertTrue(passed, msg)
+
+    def test_many_threaded(self):
+        # Fork randomly while doing operations.
+
+        clients = []
+        for _ in range(10):
+            clients.append(rs_or_single_client())
+
+        class ForkThread(ExceptionCatchingThread):
+            def __init__(self, runner, clients):
+                self.runner = runner
+                self.clients = clients
+                self.fork = False
+
+                super().__init__(target=self.fork_behavior)
+
+            def fork_behavior(self) -> None:
+                # The sequence of actions should be somewhat reproducible.
+                # If truly random, there is a chance we never actually fork.
+                # The scheduling is somewhat random, so rely upon that.
+                def action(client):
+                    client.admin.command("ping")
+                    return 0
+
+                for i in range(200):
+                    # Pick a random client.
+                    rc = self.clients[i % len(self.clients)]
+                    if i % 50 == 0 and self.fork:
+                        # Fork
+                        pid = os.fork()
+                        if pid == 0:  # Child => Can we use it?
+                            code = -1
+                            try:
+                                code = action(rc)
+                            finally:
+                                os._exit(code)
+                        else:  # Parent => Child work?
+                            self.runner.assertEqual(0, os.waitpid(pid, 0)[1])
+                    action(rc)
+
+        threads = [ForkThread(self, clients) for _ in range(10)]
+        threads[-1].fork = True
+        for t in threads:
+            t.start()
+
+        for t in threads:
+            t.join()
+
+        for c in clients:
+            c.close()
```

Comment:
Better to use self.addCleanup to close the clients directly after creating them to make sure they get closed even if the test fails.